### ISSUE TITLE
Ability to skip a testcase - from within the testcase.

### ISSUE THREAD

**Issue by eyalroz** [onqtam/doctest#322]
This is a feature request.

doctest supporting skipping test suites and test cases. But - if you've started running a test case, it will either succeed or fail. I want to have a function or a macro using which the testcase's _own code_ can decide "Nope, we can't really run the test, let's mark it as skipped". This is relevant when the conditions for running a test cannot be determined at compile time, and are impossible or at least inconvenient to determine at test-program-invoke time (when we would typically decide on skipping tests).

* doctest version: **v42.42.42**
* Operating System: **Devuan GNU/Linux 3.0**
* Compiler+version: **GCC 9.2.1**

**Comment by onqtam** [comment]
I guess for now you could simply return immediately - it will be counted as a successful test case instead of a skipped one. Alternatively you could have a subcase within a test case and have that subcase guarded by an if statement. This is a pretty niche use case and I'm not sure when (or if) I'll get to this...

**Comment by eyalroz** [comment]
@onqtam : Indeed, I can arrange for the test to succeed. Still, I think this is not that small of a niche. If your test depends on the hardware having a certain capability, and the user building your package runs `make test`, no tests are skipped by invocation; and it's still a good idea to show which tests actually ran and which decided not to do anything.

I fully understand that this is not a high priority though...

**Comment by onqtam** [comment]
> inconvenient to determine at test-program-invoke time
What do you mean by that - using filters (through the command-line or adding them programmatically in your ```main()```) to skip a test case? Are you aware of [```doctest::skip(<precondition>)```](https://github.com/onqtam/doctest/blob/master/doc/markdown/testcases.md#decorators)? (I just wan't to make sure) Do you think what you need might be accomplished with a precondition in that decorator? Note that whatever the precondition is, the code will be executed while the program is starting and globals are initialized - so before entering ```main()```.

**Comment by eyalroz** [comment]
> Are you aware of doctest::skip(<precondition>)?

Actually, I was not. But - like you've said, it's static-initialization code. And I would need code that may be test-case-specific - which I would really rather not put in `main()`.

**Comment by cdeln** [comment]
> This is relevant when the conditions for running a test cannot be determined at compile time, and are impossible or at least inconvenient to determine at test-program-invoke time (when we would typically decide on skipping tests).

@eyalroz This sounds like it's impossible to determine at both compile time AND runtime (or did my morning coffee just not kick in yet?)

I think this is a typical "configure time" thing. When you configure you project, you can perform platform specific testing using the classical "check has this feature" by compiling / running a small test program. As your feature tests pass/fail you can add definitions such that your test code knows what hardware is available at compile time. This is a well supported workflow in cmake.

If you still think it is a good idea to add another doctest macro, please provide an example of how that would look like so that I understand better and we can discuss it.

**Comment by onqtam** [comment]
I think he meant something like a runtime ```doctest::abort_as_skipped()``` function with a predicate that's evaluated after main() has been entered - not during the initialization of globals (when tests are registered and the skip() decorator is evaluated). Probably doable, maybe messy.

**Comment by cdeln** [comment]
Ok, let us be a bit concrete. The feature request is something like this

```cpp
TEST_CASE("test-case") {
    if (is_supported_hardware()) {
        REQUIRE(hardware_specific_function());
    } else {
        SKIP("feature not supported by hardware");
    }
}
```

possibly merged into a single macro

```cpp
TEST_CASE("test-case") {
    REQUIRE_IF(is_supported_hardware(), hardware_specific_function());
}
```

which can be done with pre-processor if we know this at compile time

```cpp
#if HAS_SOME_HARDWARE_FEATURE
TEST_CASE("test-case") {
    REQUIRE(hardware_specific_function());
}
#endif
```

I just want to bone out what is the feature here.

**Comment by eyalroz** [comment]
> Ok, let us be a bit concrete. The feature request is something like this
> 
> ```c++
> TEST_CASE("test-case") {
>     if (is_supported_hardware()) {
>         REQUIRE(hardware_specific_function());
>     } else {
>         SKIP("feature not supported by hardware");
>     }
> }
> ```

Yes, this. I'm not bike-shedding the exact syntax (well, not in this bug anyway...) so whatever is reasonably easy to implement would do.


> 
> possibly merged into a single macro
> 
> ```c++
> TEST_CASE("test-case") {
>     REQUIRE_IF(is_supported_hardware(), hardware_specific_function());
> }
> ```

So, here we're already seeing some movement towards an assumption that maybe this could be done at compile time. It's important to point out that the ask is for the ability to skip/abort-as-skipped after some run-time work has been done, and based on its results.


> 
> which can be done with pre-processor if we know this at compile time
> 
> ```c++
> #if HAS_SOME_HARDWARE_FEATURE
> TEST_CASE("test-case") {
>     REQUIRE(hardware_specific_function());
> }
> #endif
> ```

But you cannot know at compile-time whether which machine you'll run on, nor that you'll always run on the same one (!) 

Now, we do make some compile-time assumptions at least about the CPU architecture. But we don't bolt-in a choice of, say, who is your Ethernet adapter's manufacturer.

Plus, there could be other run-time conditions:

* "If I've managed to downloaded some file off the Internet, continue the test, otherwise skip"
* "If the computer case' ambient temperature is no more than T, continue the test, otherwise skip"

etc.

**Comment by cdeln** [comment]
> Yes, this. I'm not bike-shedding the exact syntax (well, not in this bug anyway...) so whatever is reasonably easy to implement would do.

Refined the example, also with dummy implementations to make a minimal example

```cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest/doctest.h>

bool is_supported_hardware() {
    return true;
}

int hardware_specific_function() {
  return 1337;
}

TEST_CASE("test-case") {
    if (is_supported_hardware()) {
        REQUIRE(hardware_specific_function());
    } else {
        MESSAGE("feature not supported by hardware");
    }
}
```

> But you cannot know at compile-time whether which machine you'll run on, nor that you'll always run on the same one (!)

Can you elaborate? When you say "machine", what do you mean exactly? I was talking about arch/OS stuff, which you ofc do know when you compile your code. If you with "machine" mean what peripherals are plugged into your computer at runtime, then no, ofc we can't decide that at compile time.

> Plus, there could be other run-time conditions:

These examples are absolutely crazy! I am very curious, do you actually work with test environments this unstable?

I think we need to discuss if doctest is the right framework for the job when you are testing peripherals and hardware related stuff that can randomly fail. Personally I would never want to have a test suite that can pass/fail depending on external conditions. I think what you are looking for or something with a more statistical nature? Or how would you aggregate the test information after you run it, knowing that there is a mix of random failures which are just logged with a message?

If you really do test unstable stuff like you describe, is there any improvement over the `if/else` + `MESSAGE` solution outlines above that can be done?

**Comment by eyalroz** [comment]
> Can you elaborate? When you say "machine", what do you mean exactly? I was talking about arch/OS stuff, ...  If you with "machine" mean what peripherals are plugged into your computer at runtime, then no, ofc we can't decide that at compile time.

It's any feature of the physical runtime environment. Do I have my USB drive plugged in? Do I haven GPU that's new enough? Or maybe - is the ambient temperature within an acceptable range? Is it Saturday now? Do I have at least 1 TB of free disk space? etc.

> which you ofc do know when you compile your code.

Actually you don't even know that, because you can run the same code on different OS'es or OS versions, or on different CPUs, if the libraries are compatible and the CPU architectures are compatible. Look at, say, [CMake's binary distribution](https://cmake.org/download/) - a single binary for all Linux distros on all x86_64 CPUs.

**Comment by mitchgrout** [comment]
Just chiming in, easiest solution is just a `SKIP()` macro, with an optional `SKIP(expression)` form, which will more-or-less cause the test to fail like a `FAIL()` might. However, we can slap on an extra context option to say, "allow skipped tests". Basically, a skipped test _should_ produce a failure normally, but we also don't want that to block things.

I've slated this for v2.5.0.
